libs/langgraph/langgraph/prebuilt/chat_agent_executor.py

```diff
@@ -164,10 +180,73 @@ async def acall_tool(state: AgentState, config: RunnableConfig):
     return workflow.compile()
 
 
+def _get_state_modifier_runnable(state_modifier: Optional[StateModifier]) -> Runnable:
+    if state_modifier is None:
+        state_modifier_runnable = RunnableLambda(lambda state: state["messages"])
+    elif isinstance(state_modifier, str):
+        _system_message: BaseMessage = SystemMessage(content=state_modifier)
+        state_modifier_runnable = RunnableLambda(
+            lambda state: [_system_message] + state["messages"]
+        )
+    elif isinstance(state_modifier, SystemMessage):
+        state_modifier_runnable = RunnableLambda(
+            lambda state: [state_modifier] + state["messages"]
+        )
+    elif isinstance(state_modifier, types.FunctionType):
+        state_modifier_runnable = RunnableLambda(state_modifier)
+    elif isinstance(state_modifier, Runnable):
+        state_modifier_runnable = state_modifier
+    else:
+        raise ValueError(
+            f"Got unexpected type for `state_modifier`: {type(state_modifier)}"
+        )
+
+    return state_modifier_runnable
+
+
+def _convert_messages_modifier_to_state_modifier(
+    messages_modifier: MessagesModifier,
+) -> StateModifier:
+    state_modifier: StateModifier
+    if isinstance(messages_modifier, (str, SystemMessage)):
+        return messages_modifier
+    elif isinstance(messages_modifier, types.FunctionType):
+
+        def state_modifier(state: AgentState) -> Sequence[BaseMessage]:
+            return messages_modifier(state["messages"])
+
+        return state_modifier
+    elif isinstance(messages_modifier, Runnable):
+        state_modifier = (lambda state: state["messages"]) | messages_modifier
+        return state_modifier
+    raise ValueError(
+        f"Got unexpected type for `messages_modifier`: {type(messages_modifier)}"
+    )
+
+
+def _get_model_preprocessing_runnable(
+    state_modifier: Optional[StateModifier],
+    messages_modifier: Optional[MessagesModifier],
+):
+    # Add the state or message modifier, if exists
+    if state_modifier is not None and messages_modifier is not None:
+        raise ValueError(
+            "Expected value for either state_modifier or messages_modifier, got values for both"
+        )
+
+    if state_modifier is None and messages_modifier is not None:
+        state_modifier = _convert_messages_modifier_to_state_modifier(messages_modifier)
+
+    return _get_state_modifier_runnable(state_modifier)
+
+
+@deprecated_parameter("messages_modifier", "0.1.8", "state_modifier", removal="0.2.0")
 def create_react_agent(
     model: LanguageModelLike,
     tools: Union[ToolExecutor, Sequence[BaseTool]],
-    messages_modifier: Optional[Union[SystemMessage, str, Callable, Runnable]] = None,
+    agent_state: Optional[AgentState] = None,
```

Comment:
Shouldn't this be `Type[AgentState]`? Or use a TypeVar like:

```
S = TypeVar("S", bound="AgentState")
state: Optional[Type[S]] = None
```
And then update the state modifier type to accept either agent state or S? (if that's possible)

Reply:
Also personal nit: could this just be `state_schema` ? That's what it is and i'd rather it be the same as in STateGraph.

Right now, the term "state" is overloaded betwen the schema, the instance of that state, and then the general "location" the state machine is in

Reply:
Also, adding these in the current order is technically a breaking change since someone may have done `create_react_agent(model, tools, lambda messages: ...)`. Would prefer if we use `*` and make these all keyword only